fabric-registry-sync-v0/src/main/java/net/fabricmc/fabric/impl/registry/sync/packet/DirectRegistrySyncPacket.java

```diff
@@ -0,0 +1,189 @@
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import it.unimi.dsi.fastutil.ints.Int2ObjectLinkedOpenHashMap;
+import it.unimi.dsi.fastutil.ints.Int2ObjectMap;
+import it.unimi.dsi.fastutil.objects.Object2IntMap;
+import org.jetbrains.annotations.Nullable;
+
+import net.minecraft.network.PacketByteBuf;
+import net.minecraft.util.Identifier;
+
+import net.fabricmc.fabric.impl.registry.sync.map.IdMap;
+import net.fabricmc.fabric.impl.registry.sync.map.RegistryMap;
+
+/**
+ * A more optimized method to sync registry ids to client.
+ * Produce smaller packet than old {@link NbtRegistrySyncPacket nbt-based} method.
+ *
+ * <p>This method optimize the packet in multiple way:
+ * <ul>
+ *     <li>Directly write into the buffer instead of using an nbt;</li>
+ *     <li>Group all {@link Identifier} with same namespace together and only send those unique namespaces once for each group;</li>
+ *     <li>Group adjacent rawIds together and only send the difference of the first rawId and the last rawId of the bulk before.
+ *     This is based on the assumption that mods generally register all of their object at once,
+ *     therefore making the rawIds somewhat densely packed.</li>
+ * </ul>
+ */
+public class DirectRegistrySyncPacket implements RegistrySyncPacket {
+	public static final Identifier ID = new Identifier("fabric", "registry/sync/direct");
+
+	private static final DirectRegistrySyncPacket INSTANCE = new DirectRegistrySyncPacket();
+
+	private DirectRegistrySyncPacket() {
+	}
+
+	public static DirectRegistrySyncPacket getInstance() {
+		return INSTANCE;
+	}
+
+	@Override
+	public Identifier getPacketId() {
+		return ID;
+	}
+
+	@Override
+	public void writeBuffer(PacketByteBuf buf, RegistryMap map) {
+		// Group registry ids with same namespace
+		Map<String, List<Identifier>> regNamespaceGroups = map.keySet().stream()
+				.collect(Collectors.groupingBy(Identifier::getNamespace));
+
+		buf.writeVarInt(regNamespaceGroups.size());
+
+		regNamespaceGroups.forEach((regNamespace, regIds) -> {
+			buf.writeString(regNamespace);
+			buf.writeVarInt(regIds.size());
+
+			for (Identifier regId : regIds) {
+				buf.writeString(regId.getPath());
+
+				IdMap idMap = map.get(regId);
+
+				// Group object ids with name namespace
+				Map<String, Int2ObjectMap<List<String>>> idNamespaceGroups = new LinkedHashMap<>();
+
+				for (Object2IntMap.Entry<Identifier> idPair : idMap.sortedEntryList()) {
+					Identifier id = idPair.getKey();
+					int rawId = idPair.getIntValue();
+
+					Int2ObjectMap<List<String>> adjacentRawIds = idNamespaceGroups.computeIfAbsent(getNamespace(id), s -> new Int2ObjectLinkedOpenHashMap<>());
+
+					// Group adjacent rawIds together
+					List<String> rawIdGroup;
+
+					if (adjacentRawIds.containsKey(rawId - 1)) {
+						rawIdGroup = adjacentRawIds.remove(rawId - 1);
+					} else {
+						rawIdGroup = new ArrayList<>();
+					}
+
+					rawIdGroup.add(id.getPath());
+					adjacentRawIds.put(rawId, rawIdGroup);
+				}
```

Comment:
I see a few ways the producing part of the loop that iterates over the registries could be made much more efficient.

Grouping by namespace into a sorted list of entries in the namespace first would allow you to use a more efficient strategy to find consecutive ranges of raw IDs. See [this](https://stackoverflow.com/a/55914488) for some inspiration. It could represent each range as a slice of the sorted list. It would also limit what needs to be in memory to one namespace at a time.

Reply:
Changed, how about now?